libvuln/updates/manager.go

```diff
@@ -0,0 +1,256 @@
+	interval time.Duration
+	// configs provided to updaters once constructed.
+	configs Configs
+	// provided to construct distributed locks.
+	// TODO(louis): this will be replaced by newest distlock implementation.
+	pool   *pgxpool.Pool
+	client *http.Client
+	store  vulnstore.Updater
+}
+
+// NewManager will return a manager ready to have its Start or Run methods called.
+func NewManager(ctx context.Context, store vulnstore.Updater, pool *pgxpool.Pool, outOfTree []driver.Updater, client *http.Client, opts ...ManagerOption) (*Manager, error) {
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	// the default Manager
+	m := &Manager{
+		store:     store,
+		pool:      pool,
+		factories: updater.Registered(),
+		batchSize: DefaultBatchSize,
+		interval:  DefaultInterval,
+		client:    client,
+	}
+
+	// out-of-tree updaters is a required dependency
+	// even tho it could be empty.
+	//
+	// this avoids a dependecy issue which would force ManagerOption(s)
+	// to be ran in a specific order.
+	if len(outOfTree) != 0 {
+		// merge determined updaters with any out-of-tree updaters
+		us := driver.NewUpdaterSet()
+		for _, u := range outOfTree {
+			if err := us.Add(u); err != nil {
+				log.Warn().Err(err).Str("updater", u.Name()).Msg("Duplicate updater, skipping")
+			}
+		}
+		m.factories["outOfTree"] = driver.StaticSet(us)
+	}
+
+	// these options can be ran order independent.
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m, nil
+}
+
+// Start will run updaters at the given interval.
+//
+// Start is designed to be ran as a go routine.
+// Cancel the provided ctx to end the updater loop.
+func (m *Manager) Start(ctx context.Context) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "libvuln/updates/Manager.Start").
+		Logger()
+	ctx = log.WithContext(ctx)
+	if m.interval == 0 {
+		return fmt.Errorf("Manager must be configured with an interval to start")
+	}
+	log.Info().Str("interval", m.interval.String()).Msg("Starting background updaters")
+
+	t := time.NewTicker(m.interval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+			m.Run(ctx)
+		}
+	}
+}
+
+// Run constructs updaters from factories, configures them
+// and runs them in Manager.batchSize batches.
+//
+// Run is safe to call at anytime, regardless of whether
+// background updaters are running.
+func (m *Manager) Run(ctx context.Context) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "libvuln/updates/Manager.Run").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	updaters := []driver.Updater{}
+	// constructing updater sets may require network access,
+	// depending on the factory.
+	// if construction fails we will simply ignore those updater
+	// sets.
+	for _, factory := range m.factories {
+		set, err := factory.UpdaterSet(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("Error while calling factory for updater set. Updaters in this factory will be excluded from batch.")
```

Comment:
```suggestion
			log.Error().Err(err).Msg("excluding facotry")
```

The log message already indicates there's an error.